state.go

```diff
@@ -397,6 +397,16 @@ func NewUpdateErrorState(err menderError, update UpdateResponse) State {
 func (ue *UpdateErrorState) Handle(c Controller) (State, bool) {
 	// TODO error handling
 	c.ReportUpdateStatus(ue.update, statusFailure)
+
+	// TODO fetch logs from logger
+	c.UploadLog(ue.update, []LogEntry{
+		LogEntry{
+			Timestamp: time.Now().Format(time.RFC3339),
+			Level:     "error",
+			Message:   "update failed",
+		},
+	})
```

Comment:
> UpdateErrorState should never be reached in a daemon running with a non-committed update, because being in that situation means that there will be an immediate reboot and most likely rollback.

IMO any update state (fetch/install/commit) can transition to `UpdateErrorState`. What happens next should depend on where in the update pipeline the problem occurred. Fetch/install is harmless so we could go back to `UpdateCheck`. In contrast, commit failure could potentially trigger a rollback (something we still need to implement). Each update error may require additional reporting and upload of logs. That's also a reason why update information is captured in update related states as data (incl. `UpdateErrorState`). This is different from `ErrorState` that is suitable for other generic errors (no that we have many of those) and requires no special reporting.

I agree with your other points. There's only one concern, we need to preserve the state across reboots. Maybe that's something worth working on next.